agent_patch_distr.py

```diff
@@ -0,0 +1,245 @@
+    return
+
+def loop_batches_small_patch(agent_or_patch, folderpath, ENV_HEIGHT, ENV_WIDTH, posx, posy, RADIUS_AGENT, N, N_RESOURCES, RADIUS_RESOURCE, num_batches, window_pad):
+    x_values = []
+    y_values = []
+    for j in range(0,num_batches):
+        # TODO check again: see x = np.random.randint(float, float) in sims.py. For float it rounds down so here we need to ceil
+        x_values = np.append(x_values, posx[j,:,0][:int(N)] - window_pad + int(RADIUS_AGENT))
+        y_values = np.append(y_values, posy[j,:,0][:int(N)] - window_pad + int(RADIUS_AGENT))
+
+    # making sure that no drawn location is outside the arena borders
+    max_x_values = np.max(x_values)
+    max_y_values = np.max(y_values)
+    if (max_x_values > ENV_WIDTH) | (max_y_values > ENV_HEIGHT):
+        sys.exit('At least one patch location is outside the arena borders.')
+
+    # loop through agent/patch locations and "create" a filled CIRCLE for each location in pixel array
+    pixel_array = np.zeros([ENV_HEIGHT,ENV_WIDTH])
+    for k in range(0,len(x_values)):
+        pixel_array = modifiy_pixel_array_circle(pixel_array, x_values[k], y_values[k], RADIUS_AGENT)
+    # normalized_pixel_array = pixel_array / pixel_array.sum() # this is the probability density for num_batches initializations
+    normalized_pixel_array = pixel_array / num_batches # this is the frequency a pixel is covered
+    # save pixel array for adapting plots (computation of pixel array is slow)
+    np.savez(folderpath +'/{}_normalized_pixel_array_circles_R_{}_N_{}_R_R_{}_N_R_{}_batches_{}.npz'.format(agent_or_patch,round(RADIUS_AGENT,2),int(N),round(RADIUS_RESOURCE,2),int(N_RESOURCES),num_batches),normalized_pixel_array=normalized_pixel_array)
+    plot_circles(agent_or_patch, normalized_pixel_array, RADIUS_AGENT, N, RADIUS_RESOURCE, N_RESOURCES, num_batches, folderpath)
+
+    # loop through agent/patch locations and add + 1 for each PIXEL location in pixel array
+    pixel_array = np.zeros([ENV_HEIGHT,ENV_WIDTH])
+    for k in range(0,len(x_values)):
+        pixel_array[int(y_values[k]),int(x_values[k])] = pixel_array[int(y_values[k]),int(x_values[k])] + 1
+    # normalized_pixel_array = pixel_array / pixel_array.sum() # this is the probability density for num_batches initializations
+    normalized_pixel_array = pixel_array / num_batches # this is the frequency a pixel is covered
+    # save pixel array for adapting plots (computation of pixel array is slow)
+    np.savez(folderpath +'/{}_normalized_pixel_array_locations_R_{}_N_{}_R_R_{}_N_R_{}_batches_{}.npz'.format(agent_or_patch,round(RADIUS_AGENT,2),int(N),round(RADIUS_RESOURCE,2),int(N_RESOURCES),num_batches),normalized_pixel_array=normalized_pixel_array)
+    plot_centers(agent_or_patch,normalized_pixel_array, RADIUS_AGENT, N, RADIUS_RESOURCE, N_RESOURCES, num_batches, folderpath)
+    return
+
+
```

Comment:
So the above 2 methods seem to be copy/paste functions. Can you pinpoint what is the main difference between them?

Reply:
The method loop_params_batches is for experiments where the number of patches and the radii of the patches have been varied and it can be used for generating the plots for both, the agents and the patches. The method loop_batches_small_patch takes a fixed set of parameters. It is made for plotting the agents for the case of only one resource patch af radius 1. Thus, the shape of the data that contains the positions is different from the shape of data that is used in loop_params_batches. Avoiding code duplification would probably mean more nested if-branches making the code less clear. I reduced code duplification by introducing the save_pixel_array method. In pricipal you are right, this is code duplication. I just don't see an easy way to avoid this at this point.